content/blog/2019-12-18-wasm-to-evm.md

```diff
@@ -0,0 +1,83 @@
+Unfortunately, the gas price for Ewasm instructions is not out yet, we cannot compare the total cost for running an Ewasm program and a translated EVM program, but we can still optimize EVM bytecode to reduce the gas price of a function.
+
+In Wasm, function parameters are passed by local variables. They are accessed by `get_local` instruction followed by an index number. Similarly, function parameters are stored in memory, which can be accessed by memory load operation `MLOAD` with an offset. For example, in the above function with two uint256 parameters, the first parameter is at 0x04, the second parameter is at 0x24. 
+
+Wasm compiler tends to use the local variable like registers, the compiled bytecode will load data by `get_local` whenever it needs it. However, this approach is not optimal for EVM, especially when the program needs to load a parameter multiple times in a short period. 
+
+For example, `if (a > b) return a - b` needs to load both "a" and "b" twice, 
+one for `a > b`, and one for `a - b`. In EVM, it can load both parameters for the first time, so both parameters are currently first and second elements on the stack. It duplicates both parameters by calling `DUP2` twice, comparing to loading the data from memory, this approach has a lower cost. 
+```
+| |         | |         | |
+| |         | |         |a|
+| | -DUP2-> |b| -DUP2-> |b|
+|a|         |a|         |a|
+|b|         |b|         |b|
+```
+EVM provides a set of `DUP` instructions, it can duplicate one of 1~16th stack item and push it on the top of the stack. 
+
+Currently, the binary translator can translate simple Wasm functions to EVM bytecode. 
+
+While Wasm code can be easily run locally with NodeJS, I did not figure out how to run EVM code locally for running a benchmark.
+
+## Some Thoughts
+
+After working on this project, I found it is possible to translate Wasm programs to EVM bytecode to provide "forward compatibility" for Ethereum. At least, it can provide partial "forward compatibility" so that developers can write some core computation functions in languages like C++, compile to Wasm, and translate to EVM. In this way, the developer can write the function in languages they are familiar with and utilize existing compiler optimizations of those popular languages. 
+
+However, I found binary translation might not be an optimal way to do this. Even there are many similar designs in both EVM and Wasm, it cannot achieve an optimal code without understanding the higher-level meaning of the code. Therefore, I think using intermediate representation (IR) might be a better solution, as the ongoing LLVM-EVM project. The Ethereum community is also developing an Ethereum  IR call Yul, which will be able to compile both EVM and Ewasm code.  While IR can be compiled to different backend bytecode, it can preserve more information for doing optimizations. 
```

Comment:
Hmm; I'm not entirely sure why you've concluded this. It does seem like there is some potential for recovering optimizations when doing the translation, but I don't see the evidence that it's not possible—or not desirable—to do it on the Wasm representation. You could construct CFGs and DFGs and do proper analysis directly on Wasm, just like you would to LLVM IR—nothing stops you from doing that, just as we did to Bril!

Is your argument more about the "level of abstraction" difference between LLVM and Wasm, or is it about the available analysis utilities for LLVM that are absent for Wasm?